akka-stream/src/main/scala/akka/stream/scaladsl/FlowGraph.scala

```diff
@@ -1487,3 +1498,40 @@ object FlowGraphImplicits {
       builder.addEdge(source, flow, sink)
   }
 }
+
+trait Ports {
+  private var inputs: List[UndefinedSource[_]] = Nil
+  private var outputs: List[UndefinedSink[_]] = Nil
+
+  // TODO draft - could simply be also defined on Undefined elements
+  implicit class DefinableSink[T](undefined: UndefinedSink[T]) {
+    def apply(s: Sink[T])(implicit builder: FlowGraphBuilder) = builder.attachSink(undefined, s) // TODO `apply` or `attach`?
```

Comment:
If you take the Ports only as an argument to (Partial)FlowGraph then you don't need this, you can just populate the builder from the factory method. See my previous comment.